Salvage #13772

Swiftb0y opened the pull request:
Simplified alternative of #13772 which implements the original scope I proposed.

CC @christophehenry @JoergAtGithub

christophehenry committed:
Expose convertCharset convenience function to controller

Allow to convert from UTF-8 to whatever encoding the device supports

Swiftb0y committed:
simplify ControllerScriptInterfaceLegacy::convertCharset

acolombier commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -658,6 +661,78 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+    case ISO_10646_UCS_2:
+        return 68;
+    }
+    // unreachable (TODO assert false?)
```
That assert would make sense IMO

Swiftb0y replied:
yeah, still TODO

acolombier reviewed (approved):
LGTM!

Swiftb0y committed:
feat: add a couple more encodings to enum, don't throw on invalid codec

Swiftb0y committed:
fix: don't convert invalid chars to null

Returning part of a usable result instead of nullbytes since
those likely terminate the string early or even corrupt the
underlying binary message format the buffer is embedded in.

JoergAtGithub commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:
```diff
@@ -72,6 +86,13 @@ class ControllerScriptInterfaceLegacy : public QObject {
+                    targetCharset,
+            const QString& value);
+
+    Q_INVOKABLE QByteArray convertCharset(const QString& targetCharset, const QString& value);
```
```suggestion
```
This API is unsafe and error prone! The number of supportet charsets must mit finite and tesable. This change was already requested in the orignial PR # 13772 and applies here as well!

Swiftb0y committed:
fix: test

Swiftb0y committed:
align closer with upstream

christophehenry commented on res/controllers/engine-api.d.ts:
```diff
@@ -301,4 +301,35 @@ declare namespace engine {
+     * @param value The string to encode
+     * @returns The converted String as an array of bytes. Will return an empty buffer on conversion error or unavailable charset.
+     */
+    function convertCharset(targetCharset: string, value: string): ArrayBuffer
```
Wait this is still documented even though it's now an internal API?

Swiftb0y commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -1052,3 +1057,45 @@ void ControllerScriptInterfaceLegacy::softStart(int deck, bool activate, double
+    const char* encoderName = encoderNameArray.constData();
+#endif
+    QStringEncoder fromUtf16 = QStringEncoder(encoderName);
+    if (!fromUtf16.isValid()) {
```
see the commit message, writing the replacement char is better than replacing with null bytes IMO.

christophehenry replied:
This may be a problem. During the tests I noticed that replacement char varies between Ubuntu and Fedora (maybe between Qt versions). Replace invalid chars with `\0x00` is the most predictable option.

Swiftb0y replied:
Are you sure? I think it depends on the codec. The Qt docs say they use [`QChar::ReplacementCharacter`](https://doc.qt.io/qt-6/qchar.html#SpecialCharacter-enum) [or a question mark](https://doc.qt.io/qt-6/qstringconverter.html#Flag-enum).

christophehenry replied:
[This is what I noticed here](https://github.com/mixxxdj/mixxx/pull/13772#issuecomment-2437566659). The relevant commit is probably [this one](https://github.com/mixxxdj/mixxx/commit/a12e5e3b1f0de67f0950dfca248da0d911132b2d).

christophehenry replied:
I'll try to reproduce later this week if I have time but I don't think this should be a blocker. We'll have time to change it later or introduce a new API if this proves unstable.

JoergAtGithub replied:
@christophehenry The flags here are the last open point, did you find to check this out?

christophehenry replied:
Oh sorry I didn't know you were waiting for me. I couldn't reproduce this. Feel free to merge as is.

JoergAtGithub replied:
Ok, than lets merge this!

christophehenry commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -658,6 +661,84 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+            QByteArrayView::fromArray({'\x48', '\x65', '\x6c', '\x6c', '\x6f', '\x21'}));
+}
+
+TEST_F(ControllerScriptEngineLegacyTest, convertCharsetCorrectValueStringCharset) {
+    const auto result = evaluate("engine.convertCharset('ISO-8859-15', 'Hello!')");
+
+    // ISO-8859-15 ecoded 'Hello!'
+    EXPECT_EQ(qjsvalue_cast<QByteArray>(result),
+            QByteArrayView::fromArray({'\x48', '\x65', '\x6c', '\x6c', '\x6f', '\x21'}));
+}
+
+TEST_F(ControllerScriptEngineLegacyTest, convertCharsetUnsupportedChars) {
+    auto result = qjsvalue_cast<QByteArray>(
+            evaluate("engine.convertCharset('ISO-8859-15', 'مايأ نامز')"));
+    char sub = '\x1A'; // ASCII/Latin9 SUB character
+    EXPECT_EQ(result,
+            QByteArrayView::fromArray(
+                    {sub, sub, sub, sub, '\x20', sub, sub, sub, sub}));
+}
```
Those tests aren't relevant anymore since it's an internal API now, are they?

Swiftb0y replied:
mostly yes. I should translate them to their `enum` equivalent so we can at least be reasonably sure about the output data.

Swiftb0y replied:
I changed them. please take a look

christophehenry commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -658,6 +661,84 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
     EXPECT_DOUBLE_EQ(1.0, pass->get());
 }
 
+TEST_F(ControllerScriptEngineLegacyTest, convertCharsetUndefinedOnUnknownCharset) {
+    const auto result = evaluate("engine.convertCharset('NULL', 'Hello!')");
+
+    EXPECT_EQ(qjsvalue_cast<QByteArray>(result), QByteArrayView(""));
+}
```
[Ditto](https://github.com/mixxxdj/mixxx/pull/13935/files#r1862222749).

Swiftb0y replied:
yeah... Not quite sure what to do here honestly. It seems that there are implicit conversions to enums happening, which isn't great if it just silently succeeds. I wonder what value plain `undefined` results in? I'd guess `undefined`->`0`->`US_ASCII`?

Swiftb0y replied:
I just removed it for now. I don't really know how what the proper way to handle implicit enum conversions would be. We could take a QJSValue instead and implement our own checks, but ideally Qt would warn by itself.

Swiftb0y committed:
feat: add more sensible WellKnownCharsets

christophehenry commented on res/controllers/engine-api.d.ts:
```diff
@@ -301,4 +301,50 @@ declare namespace engine {
+     * @param value The string to encode
+     * @returns The converted String as an array of bytes. Will return an empty buffer on conversion error or unavailable charset.
+     */
+    function convertCharset(targetCharset: string, value: string): ArrayBuffer
```
Don't forget to remove this version.

Swiftb0y committed:
refactor: replace irrelevant test with more meaningfull

Swiftb0y committed:
fix: remove private API doc

christophehenry commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -1052,3 +1057,79 @@ void ControllerScriptInterfaceLegacy::softStart(int deck, bool activate, double
+
+QByteArray ControllerScriptInterfaceLegacy::convertCharsetInternal(
+        const QString& targetCharset, const QString& value) {
+#if QT_VERSION >= QT_VERSION_CHECK(6, 8, 0)
```
The 6.4 version has disappeared?

Swiftb0y replied:
yeah, in hindsight I there was no reason to use `QTextEncoder` when QStringEncoder is also available, so we dropped it. QStringEncoder is also available in versions older than 6.4, so we can simply use that there too.

christophehenry commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -1,5 +1,10 @@
 #include "controllerscriptinterfacelegacy.h"
 
+#if QT_VERSION < QT_VERSION_CHECK(6, 4, 0)
+#include <QTextCodec>
+#else
+#include <QStringEncoder>
+#endif
```
```suggestion
#include <QStringEncoder>
```

Swiftb0y committed:
fix: remove unnecessary include

Swiftb0y commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -1,5 +1,10 @@
+#include <QTextCodec>
+#else
+#include <QStringEncoder>
+#endif
```
done

JoergAtGithub commented on res/controllers/engine-api.d.ts:
```diff
@@ -301,4 +301,41 @@ declare namespace engine {
      */
     function softStart(deck: number, activate: boolean, factor?: number): void;
+
+    enum WellKnownCharsets {
```
```suggestion
    enum Charsets {
```

JoergAtGithub replied:
WellKnownCharsets still appears here

Swiftb0y replied:
whoops yeah, I only `grep` in `src/` and forgot to double check. Fixed and amended.

Swiftb0y committed:
refactor: rename `WellKnownCharsets` to `Charset`

Swiftb0y commented:
@JoergAtGithub Anything left do to here? Pre-commit failure is unrelated. The commit history is garbage so I'd like to squash everything into a single commit.
